### PR TITLE
feat!: StreamId

### PR DESCRIPTION
**Pull request by bartelink**
Heavily inspired by #99 and related out of band discussions with @nordfjord 

Intended to replace Equinox.StreamId, Equinox.Core.StreamId and Equinox.Core.StreamName

### CODE REVIEW

File: src/FsCodec/StreamName.fs [hunk 1]
```diff
@@ -15,116 +15,88 @@ and [<Measure>] streamName
...
-    /// Recommended way to specify a stream identifier; a category identifier and an streamId representing the aggregate's identity
-    /// category is separated from id by `-`
-    let create (category : string) (streamId : string) : StreamName =
-        Internal.ofCategoryAndStreamId (category, streamId) |> UMX.tag
+        /// <summary>Attempts to split a Stream Name in the form <c>{category}-{streamId}</c> into its two elements.
+        /// The <c>{streamId}</c> segment is permitted to include embedded '-' (dash) characters
+        /// Returns <c>None</c> if it does not adhere to that form.</summary>
+        let tryParse (raw: string): struct (string * StreamId) voption =
+            match raw.Split(Category.separator, 2) with
+            | [| cat; id |] -> ValueSome struct (cat, StreamId.Elements.trust id)
```
Comment (nordfjord) [inline src/FsCodec/StreamName.fs:49]: id is not guaranteed to contain no `_` at this point

Comment (nordfjord) [reply src/FsCodec/StreamName.fs:49]:
```
> "a-b_c".Split("-", 5);;
val it: string array = [|"a"; "b_c"|]
```

Comment (nordfjord) [reply src/FsCodec/StreamName.fs:49]: Oh I see where you're going, this is expected. StreamName is just dealing with the `-` split, you have another layer `StreamId` dealing with the rest :+1:

**Review by nordfjord — APPROVED** [review]
Yep, this makes sense!

File: src/FsCodec/StreamName.fs (hunk 1 above)
Comment (bartelink) [reply src/FsCodec/StreamName.fs:49]: we guard against embedded `-` chars in the cat on creation
but the bottom line is anything up to the first `-` is the cat. The rest is the streamid
And we are returning it as a streamid, which is allowed to contain `_`s (but normal composition ensures you wont be using embedded `_` within them
Maybe I'm not getting your point ;)

Comment (bartelink) [reply src/FsCodec/StreamName.fs:49]: ha, sorry, missed that comment, bad wifi...

Do keep looking though - I'll validate this with Equinox and Propulsion but the next steps are:
- see in the context of an app
- see what can be learned from levelling with the JS
- get the tests more complete
there's def potential to improve naming and/or simplify here

File: src/FsCodec/StreamName.fs [hunk 2]
```diff
@@ -15,116 +15,88 @@ and [<Measure>] streamName
...
     // Validation helpers, etc.
-    module Internal =
+    module Category =
 
-        /// Throws if a candidate category includes a '-', is null, or is empty
-        let inline validateCategory (rawCategory : string) =
-            if rawCategory |> System.String.IsNullOrEmpty then invalidArg "rawCategory" "may not be null or empty"
-            if rawCategory.IndexOf '-' <> -1 then invalidArg "rawCategory" "may not contain embedded '-' symbols"
+        let [<Literal>] Separator = '_' // separates {category}-{streamId}
+        let [<Literal>] SeparatorStr = "_"
```
Comment (nordfjord) [inline src/FsCodec/StreamName.fs:26]:
```suggestion
        let [<Literal>] Separator = '-' // separates {category}-{streamId}
        let [<Literal>] SeparatorStr = "-"
```

Think this is the Id separator

Comment (bartelink) [reply src/FsCodec/StreamName.fs]: Good catch, thanks (maybe I need some more test coverage ;))

File: src/FsCodec/StreamName.fs [hunk 3]
```diff
@@ -15,116 +15,88 @@ and [<Measure>] streamName
...
-        | ValueNone -> NotCategorized
-
-    (* Rendering *)
-
-    /// Strip off the strong typing (It's recommended to pattern match as below in the general case)
-    let inline toString (streamName : StreamName) : string =
-        UMX.untag streamName
+    /// Creates a StreamName in the canonical form; a category identifier and an streamId representing the aggregate's identity
+    /// category is separated from id by `-`
+    let create (category: string) (streamId: StreamId): StreamName =
```
Comment (nordfjord) [inline src/FsCodec/StreamName.fs:70]: Will Equinox need to take a dep n FsCodec for this function now?

Comment (bartelink) [reply src/FsCodec/StreamName.fs:70]: looking at it atm - can either
- (tried so far) made Core depend on FsCodec (to get sid)
- move Category and Tracing into Equinox, making that available to anyone that depends only on Equinox (maybe splitting into a Stream.fs and Category.fs)

Atm I'm thinking that a StreamName.render will live in Core, and the stores will continue to work with just unadorned strings

Comment (bartelink) [reply src/FsCodec/StreamName.fs:70]: So Equinox has a dep on FsCodec now as Category is there and it needs StreamId
Dynamo and Cosmos also use the `category` function in metrics impl, but other than that, there's little manipulation of SN's within the stores - they just get it as a bare string

**Pull request merged**